### PR TITLE
Fix issue with CASE/IIF pre-calculating results

### LINKED ISSUE #49388: SQL: division by zero even if conditional function should have protected from this
`SELECT CASE WHEN languages = 0 THEN NULL ELSE (salary / languages) END FROM test_emp`

results in 

```
java.lang.ArithmeticException: / by zero
        at org.elasticsearch.xpack.sql.expression.predicate.operator.arithmetic.Arithmetics.div(Arithmetics.java:113) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.predicate.operator.arithmetic.BinaryArithmeticProcessor$NumericArithmetic.wrap(BinaryArithmeticProcessor.java:29) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.predicate.operator.arithmetic.BinaryArithmeticProcessor$BinaryArithmeticOperation.doApply(BinaryArithmeticProcessor.java:131) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.predicate.PredicateBiFunction.apply(PredicateBiFunction.java:24) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.predicate.operator.arithmetic.BinaryArithmeticProcessor.doProcess(BinaryArithmeticProcessor.java:176) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.gen.processor.BinaryProcessor.process(BinaryProcessor.java:50) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.predicate.conditional.CaseProcessor.process(CaseProcessor.java:45) ~[?:?]
        at org.elasticsearch.xpack.sql.execution.search.extractor.ComputingExtractor.extract(ComputingExtractor.java:78) ~[?:?]
        at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:128) ~[?:?]
        at org.elasticsearch.xpack.sql.execution.search.SearchHitRowSet.extractValue(SearchHitRowSet.java:31) ~[?:?]
        at org.elasticsearch.xpack.sql.execution.search.ResultRowSet.getColumn(ResultRowSet.java:37) ~[?:?]
```

### PR DESCRIPTION
**Pull request by matriv**
Previously, CaseProcessor was pre-calculating (called process())
on all the building elements of a CASE/IIF expression, not only the
conditions involved but also the results (if true/if false), as well as
the final else result. In case one of those results had an erroneous
calculation (e.g.: division by zero) this was executed and resulted in
an Exception to be thrown, even if this result was not used because of
the condition guarding it. e.g.:

```
SELECT CASE WHEN myField1 = 0 THEN NULL ELSE myField2 / myField1 END
FROM test;
```

Fixes: #49388

### CODE REVIEW

**Comment by matriv** [comment]
@astefan @costin I tried to add a unit test for this, but since in this we have fields involved (not only constants) I couldn't come up with one. Any ideas are welcome.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/CaseProcessor.java [hunk 1]
```diff
@@ -40,11 +39,13 @@ public void writeTo(StreamOutput out) throws IOException {
 
     @Override
     public Object process(Object input) {
-        List<Object> objects = new ArrayList<>(processors.size());
-        for (Processor processor : processors) {
-            objects.add(processor.process(input));
+        for (int i = 0; i < processors.size() - 2; i += 2) {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/CaseProcessor.java]: So, here you basically take a condition and you evaluate it, and if it holds then you evaluate its results and break out of the loop. If it doesn't hold (evaluates to `false`), then move on to the next set of [condition, result] and repeat the process. Can you add a comment about this, please?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/CaseProcessor.java [hunk 2]
```diff
@@ -40,11 +39,13 @@ public void writeTo(StreamOutput out) throws IOException {
...
+            if (processors.get(i).process(input)==Boolean.TRUE) {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/CaseProcessor.java]: `processors.get(i).process(input) == Boolean.TRUE`

**Review by astefan — APPROVED** [review]
LGTM. Left two cosmetic comments and  in terms of tests, I would like to see an IT one with more than one `WHEN` branch.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/CaseProcessor.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/CaseProcessor.java]: Added.

**Review by costin — APPROVED** [review]
LGTM

**Pull request merged**